alibuild_helpers/build.py

```diff
@@ -344,33 +344,30 @@ def doBuild(args, parser):
                   "  git pull --rebase\n",
                   pwd=os.getcwd(), star=star()))
 
-  # Prefetch git heads in parallel
-  debug("Prefetching git heads in parallel:")
-
-  fetchers = list(filter(lambda p: "source" in specs[p], buildOrder))
-  poolConfig = {"processes": min(max(len(fetchers), 1), 16)}
-  pool = ThreadPool(**poolConfig)
-  debug("Created thread pool with config: %s" % poolConfig)
-
-  git_commands = {} # store futures
-  for p in fetchers:
-    # Replace source with local one if we are in development mode.
+  # Clone/update repos
+  reposWithSource = list(filter(lambda p: "source" in specs[p], buildOrder))
+  if args.fetchRepos:
+    # Force update of reference repos
+    for p in reposWithSource:
+      updateReferenceRepos(args.referenceSources, p, specs[p])
+  else:
+    # Only create reference if it does not exist
+    for p in reposWithSource:
+      specs[p]["reference"] = "%s/%s" % (abspath(args.referenceSources), p.lower())
+      if not exists(specs[p]["reference"]):
+        updateReferenceRepos(args.referenceSources, p, specs[p])
+
+  # Retrieve git heads
+  git_head_fetchers = list(filter(lambda p: "reference" in specs[p], buildOrder))
+  for p in git_head_fetchers:
+    cmd = "git ls-remote --heads %s" % specs[p]["reference"]
     if specs[p]["package"] in develPkgs:
       specs[p]["source"] = join(os.getcwd(), specs[p]["package"])
-
-    cmd = "git ls-remote --heads %s" % specs[p]["source"]
-    debug("Enqueued '%s'" % cmd)
-    git_commands[p] = pool.apply_async(getstatusoutput, (cmd,))
-
-  results = {}
-  for p, future in git_commands.items():
-    results[p] = future.get()
-
-  for p, result in results.items():
-    dieOnError(result[0], "Unable to fetch from %s" % specs[p]["source"])
-    specs[p]["git_heads"] = result[1].split("\n")
-
-  debug("Finished prefetching git heads")
+      cmd = "git ls-remote --heads %s" % specs[p]["source"]
+    res, output = getStatusOutputBash(cmd)
+    output = to_unicode(output)
```

Comment:
I believe output of `git ls-remote` is ASCII, I think we don't need to convert it to Unicode. Besides, `getStatusOutputBash()` already returns Unicode, so this conversion can be dropped.

Reply:
I had problems with the `Vc` repo here. Seems the output can be non ASCII. But even better, if `getStatusOutputBash()` takes care of this :)

Reply:
Damn, I have tried it and you are right!

```
git ls-remote --heads https://github.com/VcDevel/Vc.git | iconv -futf-8 -tascii
```

ends with:

```
667f4f7bd1facd6783270ebf38f2a54a7e6bc524	refs/heads/mkretz/gh-102_AVX512
9a0d1481502f81bab311be6afa0d086e631f4df3	refs/heads/mkretz/gh-14iconv: illegal input sequence at position 994
```

What the hell man!